[mssql] Don't allow creating connection only when both Host and Provider/DSN are empty

agiudiceandrea opened the pull request:
## Description

Adds the possibility to create a connection when Provider/DSN parameter is specified and the Host parameter is empty.

See https://github.com/qgis/QGIS/pull/37467, https://github.com/qgis/QGIS/commit/89298373e0a0be4928f163dba6c1b187faca6e4d and https://github.com/qgis/QGIS/issues/38197#issuecomment-670849551

Fixes #38197

The issue addressed also affects 3.14 branch.

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

nyalldawson commented on src/providers/mssql/qgsmssqlnewconnection.cpp:
```diff
@@ -252,6 +253,6 @@ void QgsMssqlNewConnection::showHelp()
 {
   QListWidgetItem *item = listDatabase->currentItem();
-  bool enabled = !txtName->text().isEmpty() && !txtHost->text().isEmpty() && item;
+  bool enabled = !txtName->text().isEmpty() && !( txtService->text().isEmpty() && txtHost->text().isEmpty() ) && item;
```
Minor thing, but I think this would be more readable if we flipped the logic:

```suggestion
  bool disabled = txtName->text().isEmpty() || ( txtService->text().isEmpty() && txtHost->text().isEmpty() ) || !item;
```

agiudiceandrea replied:
Yes, indeed. Thanks for the suggestion. I've force pushed my branch with a new commit.

agiudiceandrea committed:
[mssql] Don't allow creating connection only when both Host and Provider/DSN are empty

Adds the possibility to create a connection when Provider/DSN parameter is specified and the Host parameter is empty.

nyalldawson commented:
Thanks!
